### PR TITLE
By default block service proxy to external IP addresses.

### PR DESCRIPTION
**Pull request by brendandburns**
**What this PR does / why we need it**:
Currently, the Service Proxy on the APIServer allows unrestricted access to any IP address that the APIServer machine can reach. This is likely undesirable in many cases.

Update the service proxy so that it filters Endpoints to only those that have a TargetRef that matches a known Pod.

Fixes https://github.com/kubernetes/kubernetes/issues/58761

**Release note**:
```release-note
Access to externally managed IP addresses via the kube-apiserver service proxy subresource is no longer allowed by default. This can be re-enabled via the `ServiceProxyAllowExternalIPs` feature gate, but will be disallowed completely in 1.11
```

### CODE REVIEW

File: pkg/registry/core/service/rest.go [hunk 1]
```diff
@@ -81,9 +82,13 @@ func NewStorage(registry Registry, endpoints endpoint.Registry, serviceIPs ipall
 		serviceNodePorts: serviceNodePorts,
 		proxyTransport:   proxyTransport,
 	}
+	redirectOnly := true
+	if utilfeature.DefaultFeatureGate.Enabled(features.ServiceProxyAllowExternalIPs) {
+		redirectOnly = false
```
Comment (liggitt) [inline pkg/registry/core/service/rest.go]: I'm not sure we can assume everything talking to this endpoint will follow redirects

File: pkg/registry/core/service/proxy.go [hunk 2]
```diff
@@ -71,6 +81,59 @@ func (r *ProxyREST) Connect(ctx genericapirequest.Context, id string, opts runti
...
+	}
+	svc, ok := obj.(*api.Service)
+	if !ok {
+		r.responder.Error(fmt.Errorf("unexpected type: %v", obj))
+		return
+	}
+	opts := &metainternalversion.ListOptions{
+		LabelSelector: labels.Set(svc.Spec.Selector).AsSelector(),
+	}
+	obj, err = r.pods.List(r.ctx, opts)
```
Comment (liggitt) [inline pkg/registry/core/service/proxy.go]: this doesn't take readiness/availability into account, right?

File: pkg/registry/core/service/proxy.go [hunk 3]
```diff
@@ -71,6 +81,59 @@ func (r *ProxyREST) Connect(ctx genericapirequest.Context, id string, opts runti
...
+		r.responder.Error(fmt.Errorf("unexpected type: %v", obj))
+		return
+	}
+	if len(podList.Items) == 0 {
+		r.responder.Error(fmt.Errorf("No matching pods found."))
+		return
+	}
+	ix := rand.Intn(len(podList.Items))
+	pod := podList.Items[ix]
+	path := fmt.Sprintf("/apis/v1/namespaces/%s/pods/%s/proxy%s", pod.Namespace, pod.Name, r.path)
```
Comment (liggitt) [inline pkg/registry/core/service/proxy.go]: this drops port and protocol info (see logic in service `REST#ResourceLocation`):
```
svcScheme, svcName, portStr, valid := utilnet.SplitSchemeNamePort(id)
```

**Comment by liggitt** [comment]
rather than switching to a redirect, it seems less disruptive to require the selected endpoint address to actually refer to a real pod in the same namespace with a matching IP

https://github.com/kubernetes/kubernetes/blob/85e0ed797fba34918847aa1c2e576c63a2e933dc/pkg/registry/core/service/rest.go#L435-L441

At that point, require the selected Address to have a targetRef, look up the pod, make sure it exists and has a matching IP.

**Comment by brendandburns** [comment]
@liggitt thanks for the feedback, I'll address the readiness and port/scheme comments.

Regarding redirect vs. validation, my concern is that as it is currently constructed, the proxy is enabling the end user to cloak their requests with the APIServer's IP address. That makes me fairly uncomfortable from an audit/network ACL perspective.

The only valid way to access the `/proxy` endpoint is via web sockets/spdy right shouldn't every web sockets/spdy client be able to handle a redirect?

**Comment by liggitt** [comment]
> The only valid way to access the /proxy endpoint is via web sockets/spdy right shouldn't every web sockets/spdy client be able to handle a redirect?

No, any http request is valid, and many http libraries don't follow redirects automatically.

**Comment by liggitt** [comment]
> the proxy is enabling the end user to cloak their requests with the APIServer's IP address. That makes me fairly uncomfortable from an audit/network ACL perspective.

if that is the issue, then this PR doesn't really help... it's just redirecting to the pods/proxy API endpoint, which still goes through the API server and cloaks the request with the APIServer's IP address, right?

**Comment by brendandburns** [comment]
The difference is that the pod proxy path tunnels things to the kubelet rather than communicating with the IP address directly. My concern isn't around losing the client IP address (that is going to happen either way as you suggest) but my concern is that the request is coming from the APIServer's IP address which is much more likely to have special permissions in network configuration rules.

Also, since the path to the kubelet is an HTTP call, there's a much better audit trail than making a random IP request from the APIServer's network namespace.

Basically, as currently written this code enables an arbitrary pod to appear to be traffic from the API server, that doesn't make me super comfortable.

**Comment by brendandburns** [comment]
also wrt to clients following redirects, sure, that's true of course, but if your client isn't handling redirects, you're probably already in a world of hurt (e.g. OAuth relies on redirects), so it doesn't seem that bad to me to say "if you want proxy, you need to handle redirects" Proxy isn't core functionality imho, so adding some limitations doesn't seem that bad.

**Comment by liggitt** [comment]
> The difference is that the pod proxy path tunnels things to the kubelet rather than communicating with the IP address directly. My concern isn't around losing the client IP address (that is going to happen either way as you suggest) but my concern is that the request is coming from the APIServer's IP address which is much more likely to have special permissions in network configuration rules

Actually, the service proxy subresource communicates with the IP the same way the pod proxy subresource does (see the transport it is given to use)

**Comment by brendandburns** [comment]
I'll take a look at the transport, if it is indeed tunnelling to the kubelet, then I agree that validation seems reasonable to me.

**Comment by brendandburns** [comment]
Hrm, as far as I read they both just use the default networking. I don't really like either of those...  I swear the pod proxy tunneled to kubelet *sigh* I agree it's no worse that what would happen with a redirect to the Pod, so I'll just do the Pod IP validation...

But I think we should also change the default so that we don't drive from the APIServer's IP address...

**Comment by brendandburns** [comment]
@liggitt re-factored as you suggested to just filter based on Pod TargetRef.

File: pkg/registry/core/service/rest.go [hunk 4]
```diff
@@ -419,6 +422,9 @@ func (rs *REST) ResourceLocation(ctx genericapirequest.Context, id string) (*url
 	if err != nil {
 		return nil, nil, err
 	}
+	if !utilfeature.DefaultFeatureGate.Enabled(features.ServiceProxyAllowExternalIPs) {
```
Comment (liggitt) [inline pkg/registry/core/service/rest.go]: seems like we'd want to do the check on a particular address inside the loop, after we've verified the subset even has the port we want, since it involves a remote call

Comment (brendandburns) [reply pkg/registry/core/service/rest.go]: Do you think it really matters? How many times is there an endpoint that doesn't match the port for the service?

(and is Proxy really a performance-sensitive call?)

Comment (liggitt) [reply pkg/registry/core/service/rest.go]: if you have a service backed by 400 pods (c.f.https://github.com/kubernetes/kubernetes/issues/58050 ), this would make it do 400 pod lookups as a pre-filter before selecting a single one and using it.

Comment (brendandburns) [reply pkg/registry/core/service/rest.go]: That's not the point, because you're going to pay that cost no matter what. My point was: how many services do you not pay it either way?

Comment (liggitt) [reply pkg/registry/core/service/rest.go]: If you find a suitable endpoint address first, then verify the pod as a second step, in the normal case, you only look up one pod, right?

Comment (brendandburns) [reply pkg/registry/core/service/rest.go]: Ok, I see your approach, I can restructure the code in that way, its more complicated then I was thinking...

**Comment by brendandburns** [comment]
@liggitt I updated to match your suggestions. It's not quite as clean as you thought because we need to retry if we fail, but we don't want to get stuck retrying forever.

This means that the behavior in some edge cases will be non-deterministic, but the happy path should be efficient.

File: pkg/registry/core/service/rest.go [hunk 5]
```diff
@@ -432,19 +435,55 @@ func (rs *REST) ResourceLocation(ctx genericapirequest.Context, id string) (*url
...
 		for i := range ss.Ports {
 			if ss.Ports[i].Name == portStr {
-				// Pick a random address.
-				ip := ss.Addresses[rand.Intn(len(ss.Addresses))].IP
-				port := int(ss.Ports[i].Port)
-				return &url.URL{
-					Scheme: svcScheme,
-					Host:   net.JoinHostPort(ip, strconv.Itoa(port)),
-				}, rs.proxyTransport, nil
+                                // This is a little wonky, but it's expensive to test for the presence of a Pod
```
Comment (liggitt) [inline pkg/registry/core/service/rest.go]: can pick a random starting point and iterate, wrapping around the end of the list, like `ssSeed` and `ss := &eps.Subsets[(ssSeed+ssi)%len(eps.Subsets)]` does above

File: pkg/registry/core/service/rest.go [hunk 6]
```diff
@@ -432,19 +435,55 @@ func (rs *REST) ResourceLocation(ctx genericapirequest.Context, id string) (*url
...
 		}
 	}
 	return nil, nil, errors.NewServiceUnavailable(fmt.Sprintf("no endpoints available for service %q", id))
 }
 
+func isValidAddress(ctx genericapirequest.Context, addr *api.EndpointAddress, pods *podstore.REST) error {
+	if addr.TargetRef == nil {
+		return fmt.Errorf("Address has no target ref, skipping: %v", addr)
+	}
+	nsCtx := genericapirequest.WithNamespace(ctx, addr.TargetRef.Namespace)
```
Comment (liggitt) [inline pkg/registry/core/service/rest.go]: we already have a context passed in with the right namespace, can use that (and should verify the addr.TargetRef.Namespace matches the namespace in the context... no cross-namespace pod references in an endpoint)

**Comment by liggitt** [comment]
one comment on making it deterministic, one on preventing cross-namespace pod lookups, then LGTM

**Comment by brendandburns** [comment]
@liggitt changes addressed, please take another look.

**Comment by liggitt** [comment]
check LGTM, though tests seem unhappy. we happy with the feature gate name?

**Comment by brendandburns** [comment]
tests seem to be unhappy b/c of a rebase conflict, which is weird b/c I'm rebase clean, I think it might have been a race in the testing, I'm re-running.

Name is fine with me, are there others who need to approve?

**Comment by brendandburns** [comment]
no I guess there is some kind of compile error after rebase
*sigh* will fix....

**Comment by liggitt** [comment]
> Name is fine with me, are there others who need to approve?

Not sure. Is the intent to deprecate and default off this "feature" eventually, leaving the feature gate as an escape hatch for people who still wanted this behavior during the deprecation period? If so, I wonder if we'd want the feature gate to sound more perjorative.

Just noticed the release-note isn't accurate... it's not defaulted off yet.

**Comment by brendandburns** [comment]
Yes, my plan is to eventually remove the "feature" since it is a bug imho, but I want to leave an escape hatch for at least one release for people whom it breaks.

I believe that the release note is correct.

The feature is "allow external ips" the default for features is `false` afaik, and thus the note:

By default disable access to external IP addresses from the apiserver service proxy.

Appears to be correct?

**Comment by liggitt** [comment]
> The feature is "allow external ips" the default for features is false afaik, and thus the note:
> 
> By default disable access to external IP addresses from the apiserver service proxy.
> 
> Appears to be correct?

You're right, though I think you need to add it to the defaultKubernetesFeatureGates so it can actually be set. Not sure what the designation is... sort of need a "DEPRECATED" state

**Comment by brendandburns** [comment]
@liggitt added to `defaultKubernetesFeatureGates` default false. Added a new section for 'backwards compatability' features.

ptal.

Thanks

**Comment by liggitt** [comment]
thanks

/lgtm

**Comment by mingfang** [comment]
I don't agree this is a bug and in fact it's a feature I depend on.
After you deprecate ServiceProxyAllowExternalIPs in 1.11, then what is the alternative replacement for this feature?

**Comment by Suhas-Git** [comment]
1). Does this feature block an External Service(or IP), that I am trying to access such as my database that is outside the kubernetes cluster. 

2). Also, if we are trying to protect the access of services, present in another namespace. Cannot the same result be achieved by NetworkPolicy objects? Instead of introducing the above feature in this PR ?
